### ISSUE TITLE
Cron Job executing twice on a single scheduled job

### ISSUE THREAD

**Issue by izAL17x9iZmxoo423gvY** [kelektiv/node-cron#489]
Hey so, every single time after the system runs for like 3 days, it will run into an issue whereby the cron job will simply execute twice on a single scheduled job.

**Comment by ncb000gt** [comment]
Can you share aa base code sample that replicates what you're seeing (without any extra custom code)?

**Comment by izAL17x9iZmxoo423gvY** [comment]
Yes, sure!

`var job = new CronJob('0 0 * * *', function() {
   // do something
}, null, true, 'America/New_York');
job.start();`

**Comment by ncb000gt** [comment]
Does the execution seem to always happen twice for every execution of the onTick after 3 days? Is it always 3 days?

I'm asking because I have an instance of this running myself and it doesn't perform this way and until somewhat recently I had it running for a couple months without the behavior your describing.

Is there something else that may be causing this? Have you reviewed your processes to make sure that only the one is running? Are you using a process manager to watch your processes and keep them alive (maybe there is something going on with that)?

**Comment by izAL17x9iZmxoo423gvY** [comment]
Yes, it usually starts to run twice after 3 days. And I am able to fix it by changing the time to 1 minute earlier, however, 3 days later, the same issue will happen again. I have also used other npm cron such as the cron-scheduler however, the same issue still occurs. And no, I was not using any process manager.

**Comment by csstoia** [comment]
Hi, we have a production system where all the jobs are running twice (sometimes four times), however they should run only once per a day or a week.
This script also shows that instead of running once in every 55 seconds the function runs twice with a 5 seconds delay. 

`new CronJob("*/55 * * * * *", function () {console.log('job started: ' + new Date())}, null, true);`

job started: Tue Apr 28 2020 22:15:00 GMT+0200 (CEST)
job started: Tue Apr 28 2020 22:15:55 GMT+0200 (CEST)
job started: Tue Apr 28 2020 22:16:00 GMT+0200 (CEST)
job started: Tue Apr 28 2020 22:16:55 GMT+0200 (CEST)
job started: Tue Apr 28 2020 22:17:00 GMT+0200 (CEST)
job started: Tue Apr 28 2020 22:17:55 GMT+0200 (CEST)
job started: Tue Apr 28 2020 22:18:00 GMT+0200 (CEST)

**Comment by jeffminsungkim** [comment]
`*/25 * * * * *, */35 * * * * *, */45 * * * * *, */55 * * * * *` having the same problem as @stcsaba mentioned above.

**Comment by MattCSmith** [comment]
I am also noticing this problem. 
I have a digital ocean droplet, using PM2.
I have two Crons (`00 59 23 * * *`  &  `00 59 23 * * Wed`)  

Usually speaking they both start of fine for a couple of days and then they would start running an additional time every couple days. I would usually just restart the service and it would be good for a couple days. 

However this time it started straight away, 
I am currently in 16 days of uptime and this is how the `00 59 23 * * *`  cron has fired:

25th - 1 Time
26th - 2 Times
27th - 2 Times
28th - 3 Times
29th - 3 Times
30th - 4 Times 
31st - 4 Times
1st - 4 Times
2nd - 4 Times
3rd - 5 Times
4th - 5 Times
5th - 5 Times
6th - 6 Times
7th - 6 Times
8th - 6 Times
9th - 7 Times
10th - 7 Times

**Comment by vikram1992** [comment]
Did anyone found a solution, mine is executing at-most 3 times.

**Comment by vikram1992** [comment]
Solved by running cron in fork mode rather than in cluster mode. But how can I distribute the cron task without executing the same task again in the next thread?

**Comment by tibfox** [comment]
I transfer funds with this cron job and I have to find a solution for it or I have to change my execution strategy to something wilder than this package. @vikram1992 how can I run it in "fork mode"?

**Comment by AbdlrahmanSaberAbdo** [comment]
I attempted to debug the problem to identify its cause. As suggested by @stcsaba, I tested it with a time duration of 55 seconds. After some debugging, I discovered an issue within the `_getNextDateFrom` method in the `time.js` file.

Specifically, within the first 55 seconds of the minute, everything worked fine. Let's take the initial main date as `1969-12-31T19:01:00.000`. However, after applying for a first job, the date would become `1969-12-31T19:01:55.000`. in the second job the problem arises (I mean when the date contains 55 seconds). In the `_getNextDateFrom` method, there are two variables: `start`, which represents the start date (`1969-12-31T19:01:55.000` in this case), and the date variable, which increments within the loop.

Here's the portion of the code that deals with the "seconds" case:

```
if (
	!(date.second in this.second) &&
	Object.keys(this.second).length !== 60
) {
	const expectedSecond =
		date.second === 59 && diff > 60000 ? 0 : date.second + 1;
	const expectedMinute = date.minute + (expectedSecond === 60);
	const expectedHour = date.hour + (expectedMinute === 60 ? 1 : 0);

	date = date.set({ second: expectedSecond });

	// Seconds can cause it too, imagine 21:59:59 -> 23:00:00.
	if (this._forwardDSTJump(expectedHour, expectedMinute, date)) {
		const [done, newDate] = this._findPreviousDSTJump(date);
		date = newDate;
		if (done) break;
	}

	continue;
}
```

Within this code, the seconds keep increasing until they reach 60. In this particular case, when the seconds become 60, the difference is 5 seconds. However, due to the condition `Object.keys(this.second).length !== 60`, which ensures that the seconds cannot reach 60, the code does not enter this condition anymore and breaks, returning the date. As a result, the difference is only 5 seconds. In the next job, it works as expected with a duration of "55" seconds, but the subsequent job is fired after 5 seconds, and so on.

I suspect that this issue may also occur for durations of "45", "35", and "25" seconds, or any duration that results in a floating-point number when dividing 60 by the number of seconds.

This is just the explanation for the "seconds" case, but I suspect there might be a similar behavior for other issues. I would like to contribute and fix this problem, but it would require a significant amount of time since I'm unfamiliar with the codebase. It's surprising that this issue has persisted for three years. If any of the creators are interested in scheduling a call to resolve it together, I would be happy to do so!

**Comment by intcreator** [comment]
hey thanks for looking into it. we'll probably rewrite the `_getNextDateFrom` function as part of our roadmap for the next steps on this project. if you want you can [join the Discord server](https://discord.gg/yyKns29zch) and we can discuss how it works a little more informally

**Comment by sheerlox** [comment]
> Hi, we have a production system where all the jobs are running twice (sometimes four times), however they should run only once per a day or a week. This script also shows that instead of running once in every 55 seconds the function runs twice with a 5 seconds delay.
> 
> `new CronJob("*/55 * * * * *", function () {console.log('job started: ' + new Date())}, null, true);`
> 
> job started: Tue Apr 28 2020 22:15:00 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:15:55 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:16:00 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:16:55 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:17:00 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:17:55 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:18:00 GMT+0200 (CEST)

*from @stcsaba*

> */25 * * * * *, */35 * * * * *, */45 * * * * *, */55 * * * * * having the same problem as @stcsaba mentioned above. Also, it behaves exactly the same on every minute when the value is not divisible by 60.

*from @jeffminsungkim*

This isn't an implementation issue, this is the way the standard cron is designed: `*/55` doesn't mean "run every 55 seconds", it means "run at every 55th second". You can check on [crontab.guru](https://crontab.guru/#*/55_*_*_*_*) that this is indeed the standard behavior. For more information on this, please [refer to my comment on #681](https://github.com/kelektiv/node-cron/issues/681#issuecomment-1646799318).

**Comment by sheerlox** [comment]
> However, due to the condition `Object.keys(this.second).length !== 60`, which ensures that the seconds cannot reach 60, the code does not enter this condition anymore and breaks, returning the date. 

As a side note, this line of code checks if we should run the job every second (when seconds is set to `*` or `0-59` in the expression, we will set all keys of `this.second` to `true`, for implementation see [time.js#L761-L807](https://github.com/kelektiv/node-cron/blob/main/lib/time.js#L761-L807)).

**Comment by sheerlox** [comment]
These comments not being related to the original issue, I'm hiding them to reduce the noise on this issue.

**Comment by AbdlrahmanSaberAbdo** [comment]
> > Hi, we have a production system where all the jobs are running twice (sometimes four times), however they should run only once per a day or a week. This script also shows that instead of running once in every 55 seconds the function runs twice with a 5 seconds delay.
> > `new CronJob("*/55 * * * * *", function () {console.log('job started: ' + new Date())}, null, true);`
> > job started: Tue Apr 28 2020 22:15:00 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:15:55 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:16:00 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:16:55 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:17:00 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:17:55 GMT+0200 (CEST) job started: Tue Apr 28 2020 22:18:00 GMT+0200 (CEST)
> 
> _from @stcsaba_
> 
> > */25 * * * * *, */35 * * * * *, */45 * * * * *, */55 * * * * * having the same problem as @stcsaba mentioned above. Also, it behaves exactly the same on every minute when the value is not divisible by 60.
> 
> _from @jeffminsungkim_
> 
> This isn't an implementation issue, this is the way the standard cron is designed: `*/55` doesn't mean "run every 55 seconds", it means "run at every 55th second". You can check on [crontab.guru](https://crontab.guru/#*/55_*_*_*_*) that this is indeed the standard behavior. For more information on this, please [refer to my comment on #681](https://github.com/kelektiv/node-cron/issues/681#issuecomment-1646799318).

Interesting, thanks for clarifying this. I thought that it should run every 55 seconds. now it makes sense, I think it'd be better to add this to the docs to let the user know what is the default behavior and clear any confusion.

And one question, how this is different from the main issue? I think it's related, right? Or do I miss something?

**Comment by sheerlox** [comment]
> Interesting, thanks for clarifying this. I thought that it should run every 55 seconds. now it makes sense, I think it'd be better to add this to the docs to let the user know what is the default behavior and clear any confusion.

I'm unsure if adding this kind of detail to the documentation is relevant since this library only follows the cron standard regarding expressions (with an inconsistency on numeric months, which is detailed in the doc and will be standardized in V3).
We could add one or two links to detailed posts explaining how cron expressions work (like [cron's man page](https://man7.org/linux/man-pages/man5/crontab.5.html) and [Crontab.guru examples](https://crontab.guru/examples.html)). PRs are welcome! :)

> How this is different from the main issue? I think it's related, right? Or do I miss something?

The [OP mentioned](https://github.com/kelektiv/node-cron/issues/489#issuecomment-619696752) using this cron expression: `0 0 * * *`, which doesn't use the `*/55` pattern, so this discussion about how that pattern works doesn't seem related to me. Especially since we established that it isn't a bug and is working as it should as designed by the cron standard.

**Comment by AbdlrahmanSaberAbdo** [comment]
> I'm unsure if adding this kind of detail to the documentation is relevant since this library is only an implementation of the standard cron for Node.JS (with an inconsistency on numeric months, which is detailed in the doc and will be standardized in V3).
> We could add one or two links to detailed posts explaining how cron expressions work (like [cron's man page](https://man7.org/linux/man-pages/man5/crontab.5.html) and [Crontab.guru examples](https://crontab.guru/examples.html)). PRs are welcome! :)

I think this a nice idea to add links to detailed posts.

> The https://github.com/kelektiv/node-cron/issues/489#issuecomment-619696752 using this cron expression: 0 0 * * *, which doesn't use the */55 pattern, so this discussion about how that pattern works doesn't seem related to me. Especially since we established that it isn't a bug and is working as it should as designed by the cron standard.

I think the main problem is still the same if you checked my comment [link](https://github.com/kelektiv/node-cron/issues/489#issuecomment-1634849348), specifically this
> Within this code, the seconds keep increasing until they reach 60. In this particular case, when the seconds become 60, the difference is 5 seconds. However, due to the condition Object.keys(this.second).length !== 60, which ensures that the seconds cannot reach 60, the code does not enter this condition anymore and breaks, returning the date. As a result, the difference is only 5 seconds. In the next job, it works as expected with a duration of "55" seconds, but the subsequent job is fired after 5 seconds, and so on.

I understand that the default behavior is intended to execute at every */xx time, but this will result in duplication, as I explained earlier in */55 example in this [comment](https://github.com/kelektiv/node-cron/issues/489#issuecomment-1634849348).

**Comment by sheerlox** [comment]
> I think the main problem is still the same if you checked my comment [link](https://github.com/kelektiv/node-cron/issues/489#issuecomment-1634849348), specifically this
> 
> > Within this code, the seconds keep increasing until they reach 60. In this particular case, when the seconds become 60, the difference is 5 seconds. However, due to the condition Object.keys(this.second).length !== 60, which ensures that the seconds cannot reach 60, the code does not enter this condition anymore and breaks, returning the date. As a result, the difference is only 5 seconds. In the next job, it works as expected with a duration of "55" seconds, but the subsequent job is fired after 5 seconds, and so on.
> 
> I understand that the default behavior is intended to execute at every */xx time, but this will result in duplication, as I explained earlier in */55 example in this [comment](https://github.com/kelektiv/node-cron/issues/489#issuecomment-1634849348).

As I already stated [in this comment](https://github.com/kelektiv/node-cron/issues/489#issuecomment-1678926658):

>> However, due to the condition Object.keys(this.second).length !== 60, which ensures that the seconds cannot reach 60, the code does not enter this condition anymore and breaks, returning the date.

> @AbdlrahmanSaberAbdo as a side note, this line of code checks if we should run the job every second (when seconds is set to * or 0-59 in the expression, we will set all keys of this.second to true, for implementation see [time.js#L761-L807](https://github.com/kelektiv/node-cron/blob/main/lib/time.js#L761-L807)).

From my point of view, you're trying to debug an issue using two wrong assumptions (the first one is that `*/55` should execute "every 55 seconds", and the second one is that the purpose of `Object.keys(this.second).length !== 60` is to ensure "the seconds cannot reach 60"). I already explained in detail why those assumptions are wrong, and therefore so is your analysis.

If you still think I'm missing something and there is indeed a bug, please [open a separate issue](https://github.com/kelektiv/node-cron/issues/new?assignees=&labels=%F0%9F%9B%A0+goal%3A+fix&projects=&template=bug_issue.yaml) after updating your analysis & comments in light of my explanations.

**Comment by ianfirdaus31** [comment]
Try running the scheduler using fork mode instead of cluster mode if using pm2.

**Comment by sheerlox** [comment]
> Try running the scheduler using fork mode instead of cluster mode if using pm2. 

Hey @ianfirdaus31 👋 did you encounter this issue recently?

**Comment by ianfirdaus31** [comment]
> > Try running the scheduler using fork mode instead of cluster mode if using pm2.
> 
> Hey @ianfirdaus31 👋 did you encounter this issue recently?

Nope, it was fixed by the way i explained above

**Comment by sheerlox** [comment]
Closing this since no one reported encountering this bug in more than a year.

### Using `cron` in a clustered Node.js application

This is out of the scope of this package and is directly related to your design and implementation choices.
As an example, I've had to tackle that issue recently to balance the jobs load on an application and decided to go with only two instances so that I could use `cluster.isPrimary` & `cluster.isWorker` as the value for the `start` argument, effectively distributing my jobs between the primary and the worker.
More complex use cases with more than two instances will require a lot more thinking, and we'd be glad [to discuss](https://github.com/kelektiv/node-cron/discussions) any approaches you might come up with to tackle that, and maybe come up with a standard solution we could add to the documentation.

**Comment by Sreedhar-Nambi-Sam** [comment]
![Image](https://github.com/user-attachments/assets/ea4c341a-cf1c-4e86-bc24-60c8a51813d5)

![Image](https://github.com/user-attachments/assets/dac17535-d36e-4814-afb5-b8df370ecf03) 

The cron being triggered twice. This is happening in local enviroment.

**Comment by sheerlox** [comment]
@Sreedhar-Nambi-Sam please open a new issue and provide more details about your setup. For example, are you using Node clustering, or anything that might cause this error?

**Comment by sheerlox** [comment]
I locked this issue as it is never a bug in `cron`, but most of the time arises from a misconfiguration either locally (e.g. using Node's `cluster`) or in the cloud provider (running multiple instances and expecting only one to run the cron jobs without actually configuring this behavior).

The issue tracker being dedicated to bug reports and feature requests, but feel free to join [our Discord server](https://github.com/kelektiv/node-cron?tab=readme-ov-file#-community), where we have a dedicated support channel.
